### PR TITLE
Update JSON Schema to be v1.0 compliant

### LINKED ISSUE #867: Creating new resources and validating with JSON Schema
The official JSON API Schema (as described at http://jsonapi.org/schema) explicitly expects the ID field to be sent by clients when creating new resources. This conflicts with the following example (pulled from the official documentation):

``` json
POST /photos HTTP/1.1
Content-Type: application/vnd.api+json
Accept: application/vnd.api+json

{
  "data": {
    "type": "photos",
    "attributes": {
      "title": "Ember Hamster",
      "src": "http://example.com/images/productivity.png"
    },
    "relationships": {
      "photographer": {
        "data": { "type": "people", "id": "9" }
      }
    }
  }
}
```

When I validate the above to the official JSON Schema document, validation fails. Perhaps a seperate JSON Schema document can be provided solely for creating new resources, where the ID field is optional?

### LINKED ISSUE #1522: jsonapi schema $id property
Hi,

noticed that the [jsonapi schema](https://github.com/json-api/json-api/blob/v1.1rc3/schema) does not have the $id property. just curious if that's intended design?

The reason for asking is that currently, to reference the spec in something like following, I would need to add an $id property to the jsonapi schema? 

```
  "allOf": [
    {
      "$ref": "http://jsonapi.org/schema#/definitions/success"
    },
    {
      "$ref": "#/definitions/myDomainResponse"
    }
  ],

```

### LINKED ISSUE #1554: JSON Schema for JSON:API v1.0 is incomplete or broken
Hi all. I promised to work on this, so here goes!

The `schema` file in the root of this repo is a little lacking and in some cases inaccurate.
I feel like what makes the most sense here is to make a draft PR and then talk about the changes I've made and why.

I've not fully tested it, but it should be more than enough to get discussions going!

PR to follow!

### PR DESCRIPTION
**Pull request by VGirol**
I allowed myself to continue the work started by Relequestual in the pull request #1555.
I've updated the JSON schema for version 1.0 and added many sample files to test it. In these sample files there are both valids and invalids documents.
I've also added a JSON schema which allows to validate requests for creating new resources (still for the 1.0 version).
I hope this will help.
if that's okay with you, i could do the same for version 1.1.

closes #1127
closes #1554 
closes #1522 
closes #867

### CODE REVIEW

**Comment by auvipy** [comment]
amazing work dude

**Comment by bradjones1** [comment]
Thank you for the awesome work. It seems that we could wire up a CI job to test the schema file against the fixtures you've provided using, say, [ajv](https://ajv.js.org)?

File: _schemas/1.0/schema.json
Comment (bradjones1) [inline _schemas/1.0/schema.json:584]: I believe this is incorrect. The spec gives various examples of additional link types, and the 1.1 spec (though I know that's not what this covers) includes examples such as `describedby`.

Comment (bradjones1) [reply _schemas/1.0/schema.json:584]: To be clear, I'm suggesting removing the `additionalProperties: false` rule.

File: _schemas/1.0/schema.json
Comment (bradjones1) [inline _schemas/1.0/schema.json:464]: Here as well.

**Comment by auvipy** [comment]
@dgeb @gabesullice

File: _schemas/1.0/schema.json
Comment (bradjones1) [reply _schemas/1.0/schema.json:584]: Ref for `describedby` - https://github.com/json-api/json-api/pull/1447

This makes me question myself on whether or not the 1.0 spec specifically does or does not preclude other link types. It seems vague, and doesn't use any `MUST` or `MUST NOT` language, so I think `describedby` should be valid on 1.0, but it's... vague.

Comment (bradjones1) [reply _schemas/1.0/schema.json:584]: Ack - I am actually wrong. Per the spec,

```
Unless otherwise noted, objects defined by this specification MUST NOT contain any additional members. Client and server implementations MUST ignore members not recognized by this specification.
```

Sooo this is yet another reason to really get 1.1 published, since implementations like Drupal's `jsonapi_hypermedia` add `describedby`.

Comment (VGirol) [reply _schemas/1.0/schema.json:584]: This is exactly the reason why i added the `additionalProperties: false` rule. I was not sure the links object `MUST` or `MUST NOT` contain any additional members. So, when in doubt, I preferred to refer to :
`Unless otherwise noted, objects defined by this specification MUST NOT contain any additional members. Client and server implementations MUST ignore members not recognized by this specification.`

Comment (bradjones1) [reply _schemas/1.0/schema.json:584]: Yep. That's the correct interpretation. I don't have permission to resolve this thread but feel free to do so. Thanks for your awesome work on this.

**Review by auvipy — COMMENTED** [review]
Looks great

**Comment by VGirol** [comment]
According to this PR [#1127](https://github.com/json-api/json-api/pull/1127), i've added a schema to validate requests that update a resource or relationship.

**Comment by VGirol** [comment]
Actually, the JSON schema for v1.0 uses the draft-07 version.
Do we keep this version ? Or do i need to upgrade to the latest version (2020-12) ?

**Comment by VGirol** [comment]
Once this pr is merged (and the #1603 also), we can update the reference in https://github.com/SchemaStore/schemastore (see #1011 ).

**Comment by VGirol** [comment]
I read this comment (https://github.com/json-api/json-api/issues/975#issuecomment-253420142) and I realized that by modifying the tree structure (and the schema file name), I introduced some backward breaks.
Do you think this is a big problem ?

**Comment by VGirol** [comment]
Could someone tell me what is missing for this PR to be merged ?

File: _schemas/1.0/schema.json
Comment (bradjones1) [inline _schemas/1.0/schema.json]:
```suggestion
          "$ref": "#/definitions/relationshipsForbiddenMemberName"
```

File: _schemas/1.0/schema.json
Comment (bradjones1) [inline _schemas/1.0/schema.json]:
```suggestion
          "$ref": "#/definitions/relationshipsForbiddenMemberName"
```

File: _schemas/1.0/schema.json
Comment (bradjones1) [inline _schemas/1.0/schema.json]:
```suggestion
    "relationshipsForbiddenMemberName": {
```

**Review by bradjones1 — CHANGES_REQUESTED** [review]
Same as https://github.com/json-api/json-api/pull/1603#pullrequestreview-967214179 on 1.1; typo `s/forbiden/forbidden`

**Comment by VGirol** [comment]
Oups. Shame on me.
It's corrected.

**Comment by derekwsgray** [comment]
Woah! Wish I'd found this a couple days ago - Took a while to realize the schema linked on the jsonapi.org page is not valid (blocker was relationships vs relationshipsFromRequest)
I'm pasting all of this into our openapi.json spec

This really needs to be on the website! Thank you!

**Comment by dgeb** [comment]
> This really needs to be on the website! 

@derekwsgray Agreed, this is a high priority!

@VGirol Thanks so much for your hard work on this. I'll finish reviewing this very soon.

**Comment by derekwsgray** [comment]
There are some places that don't work in OpenAPI 3.1 (thought I thought all of json schema should have worked.)
Example, the `resource` schema uses `allOf` as a sibling to `properties`, whereas OpenAPI expects `allOf` to contain that object with properties as a second child. (hope that makes sense.)

**Comment by auvipy** [comment]
> There are some places that don't work in OpenAPI 3.1 (thought I thought all of json schema should have worked.) Example, the `resource` schema uses `allOf` as a sibling to `properties`, whereas OpenAPI expects `allOf` to contain that object with properties as a second child. (hope that makes sense.)

did you check the v1.1 complaint schema PR https://github.com/json-api/json-api/pull/1603 ? would love to have some review

**Comment by derekwsgray** [comment]
> 
> did you check the v1.1 complaint schema PR #1603 ? would love to have some review

Nope, didn't realize it existed!

**Comment by auvipy** [comment]
> > This really needs to be on the website!
> 
> @derekwsgray Agreed, this is a high priority!
> 
> @VGirol Thanks so much for your hard work on this. I'll finish reviewing this very soon.

should we consider this and the v 1.1 json-schema  updates for v1.1 final release?

File: _schemas/1.0/tests/request/resource/create/invalid/relationship_with_bad_resource_identifier.json
Comment (jelhan) [inline _schemas/1.0/tests/request/resource/create/invalid/relationship_with_bad_resource_identifier.json]: Do you mind changing to a more neutral title in the tests? This might be offending to some users of JSON:API specification.

```suggestion
      "title": "Resource identifier object used for relationship data is wrong."
```

Same applies to the other cases as well.

File: _schemas/1.0/tests/request/resource/create/valid/post_resource.json
Comment (jelhan) [inline _schemas/1.0/tests/request/resource/create/valid/post_resource.json:1]: Would be great if you could add three additional scenarios for _valid_ JSON:API documents used for creating a resource:

1. Without `relationships`
2. Without `attributes`
3. With client-generated ID

File: _schemas/1.0/tests/request/resource/update/valid/patch_resource.json
Comment (jelhan) [inline _schemas/1.0/tests/request/resource/update/valid/patch_resource.json:1]: Would be great if you could add two additional scenarios for _valid_ JSON:API documents used for updating a resource:

1. Without `relationships`
2. Without `attributes`

File: _schemas/1.0/tests/response/invalid/errors/invalid_error_objects.json
Comment (jelhan) [inline _schemas/1.0/tests/response/invalid/errors/invalid_error_objects.json:1]: Would be great if you could add some information (e.g. as `detail` property) _why_ the error objects are invalid.

File: _schemas/1.0/tests/response/invalid/relationships/to_many_linkage_twice.json
Comment (jelhan) [inline _schemas/1.0/tests/response/invalid/relationships/to_many_linkage_twice.json]: I'm not sure if this is actually invalid. See #1361.

@dgeb We may need to clarify if parallel relationships are supported or not.

File: _schemas/1.0/tests/response/invalid/resource/relationship_named_id.json
Comment (jelhan) [inline _schemas/1.0/tests/response/invalid/resource/relationship_named_id.json]: I would recommend to use a neutral example here as well. This one may not be as offending as the PHP example above. But I think for a specification it would be to not make any opinionated statements about technologies, which might be used to implement it.

**Review by jelhan — CHANGES_REQUESTED** [review]
Thanks a lot for this awesome work!

I left a few minor comments on the test cases. Please have a look.

I haven't reviewed the JSON Schema itself (yet). I'm not an expert on JSON Schema. The extensive test cases helps me a lot to verify correctness of the JSON Schema. Thanks a lot for putting that much time into them!

Do you mind resetting all changes, which are not related to the JSON Schema? Especially the changes to `_format/1.1/index.md`, `examples/index.md`, `implementations/index.md`, and `recommendations/index.md`?

Could we have `/schema` as a HTTP redirect to `_schemas/1.0/schema.json`? Or could we maybe copy it over when building the website? I worry that users may rely on that URL to fetch latest version of the schema. Would be great to have it as an alias pointing to the schema of the latest stable version.

Would be great if you could add documentation how to run the tests for the JSON Schema. `CONTRIBUTING.md` seems to be a good fit for it.

Additionally we should run the tests with the schema in GitHub actions. Do you mind setting CI up?

@dgeb I think we should try to land this first and than build the schema for v1.1 on top of it. What do you think?

File: _schemas/1.0/tests/request/resource/create/invalid/relationship_with_bad_resource_identifier.json
Comment (VGirol) [reply _schemas/1.0/tests/request/resource/create/invalid/relationship_with_bad_resource_identifier.json]: I agree with you. I think this is a copy/paste of an example found in an older version of the specs.

File: _schemas/1.0/tests/response/invalid/resource/relationship_named_id.json
Comment (VGirol) [reply _schemas/1.0/tests/response/invalid/resource/relationship_named_id.json]: Same as for PHP...

Comment (VGirol) [reply _schemas/1.0/tests/response/invalid/resource/relationship_named_id.json]: As I said for PHP, it's a copy/paste of an exemple found in the specs : [https://jsonapi.org/format/#document-resource-objects](https://jsonapi.org/format/#document-resource-objects).

File: _schemas/1.0/tests/response/invalid/relationships/to_many_linkage_twice.json
Comment (VGirol) [reply _schemas/1.0/tests/response/invalid/relationships/to_many_linkage_twice.json]: I agree with you : it's not clear if parallel relationships are supported or not.
When in doubt, I preferred to be as restrictive as possible.
As soon as the situation is clarified, I will change the schema and the test files.

Comment (bradjones1) [reply _schemas/1.0/tests/response/invalid/relationships/to_many_linkage_twice.json]: As I read that issue and the spec, I think this specific example is indeed invalid; the linked issue relating to a potential `arity` member or other way to uniquely identify them would clear that up.

File: _schemas/1.0/tests/request/resource/create/invalid/relationship_with_bad_resource_identifier.json
Comment (VGirol) [reply _schemas/1.0/tests/request/resource/create/invalid/relationship_with_bad_resource_identifier.json]: Done

File: _schemas/1.0/tests/request/resource/create/valid/post_resource.json
Comment (VGirol) [reply _schemas/1.0/tests/request/resource/create/valid/post_resource.json:1]: Done

File: _schemas/1.0/tests/request/resource/update/valid/patch_resource.json
Comment (VGirol) [reply _schemas/1.0/tests/request/resource/update/valid/patch_resource.json:1]: Done

File: _schemas/1.0/tests/response/invalid/resource/relationship_named_id.json
Comment (VGirol) [reply _schemas/1.0/tests/response/invalid/resource/relationship_named_id.json]: Done

File: _schemas/1.0/tests/response/invalid/errors/invalid_error_objects.json
Comment (VGirol) [reply _schemas/1.0/tests/response/invalid/errors/invalid_error_objects.json:1]: I have added informations about the errors present in this document, but also for all the other test files. Here is an exemple for an invalid response : 
```
{
  "data": "not valid",
  "meta": {
    "errors-present-in-document": [
      {
        "detail": "Primary data **MUST** be either: 1) a single resource object, a single resource identifier object, or null, for requests that target single resource 2) an array of resource objects, an array of resource identifier objects, or an empty array ([]), for requests that target resource collections",
        "source": {
          "pointer": "/data"
        }
      }
    ]
  }
}

```

**Comment by VGirol** [comment]
> Do you mind resetting all changes, which are not related to the JSON Schema? Especially the changes to `_format/1.1/index.md`, `examples/index.md`, `implementations/index.md`, and `recommendations/index.md`?

These changes are present on the PR for version 1.1 (#1603). I don't know why they appear in this PR. I will remove them quickly.

> Could we have `/schema` as a HTTP redirect to `_schemas/1.0/schema.json`? Or could we maybe copy it over when building the website? I worry that users may rely on that URL to fetch latest version of the schema. Would be great to have it as an alias pointing to the schema of the latest stable version.
 
It's a good idea, but I'm not sure I can make it happen. I don't know how the website is built. I'll take a look soon...

> Would be great if you could add documentation how to run the tests for the JSON Schema. `CONTRIBUTING.md` seems to be a good fit for it.

Well, i've used an hand-made validator coded in PHP.
 
> Additionally we should run the tests with the schema in GitHub actions. Do you mind setting CI up?

I've never set up CI. I'm going to try...

**Comment by bradjones1** [comment]
[ajv](ajv.js.org) is a great tool for validating json-schema, it can also validate a schema against the meta-schema, as would be the case here as well. I've set up CI for something like this for proprietary/in house projects before, if you'd like the help.

**Comment by VGirol** [comment]
I've added : 

- a script to run the tests : the script looks for all available versions of the spec and runs the tests for each version
- documentation on how to run the tests

But now, @bradjones1, i will appreciate some help to integrate the script in the CI

**Comment by bradjones1** [comment]
Ahh this is so great.

I am more experienced with GitLab CI, and I'm not even sure if either of us have sufficient permission on this repo to actually do much to initially set up/change the GitHub Actions? @dgeb any thoughts on this?

AFAICT the only real CI integration right now is GitHub pages, which appears to be served behind a TLS-terminating proxy for jsonapi.org. But that doesn't require Actions, it just uses the `gh-pages` branch by convention.

File: CONTRIBUTING.md
Comment (jelhan) [inline CONTRIBUTING.md]: I think we need some additional empty lines to ease readability and not break formatting. Especially there should be an empty line before and after code blocks.

For inline code, it should only have one backtick not three.

Sadly I can not provide a suggestion because the syntax for suggestion does not seem to support code, which itself includes three backticks.

**Comment by jelhan** [comment]
> I am more experienced with GitLab CI, and I'm not even sure if either of us have sufficient permission on this repo to actually do much to initially set up/change the GitHub Actions? @dgeb any thoughts on this?

If I recall correctly, you could test the workflow on your fork. To be run for this project, we would need to approve first. So that would slow down development.

I'm also more familiar with writing GitLab CI than GitHub Actions. But the GitHub Actions documentation looks very good. For this use case small customization to the [example included in getting started section](https://docs.github.com/en/actions/learn-github-actions/understanding-github-actions#create-an-example-workflow) might be enough. Also [Ember project provides a GitHub workflow for addons](https://github.com/ember-cli/ember-addon-output/blob/f9e2d2cd41252477c9070718212df598c495f6f2/.github/workflows/ci.yml), which has similar requirements.

Putting both together, a GitHub Actions workflow to run tests for JSON Schema might look like this:

```yml
name: tests

on:
  push:
    branches:
      - gh-pages
  pull_request: {}

jobs:
  test-json-schema:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - name: Install Node
        uses: actions/setup-node@v3
        with:
          node-version: '16'
      - name: Install dependencies
        run: npm ci
      - name: Run tests
        run:  node ./_schemas/scripts/validator.js
```

We could optimize later to only run if any file related to JSON Schema or its tests is changing.

Disclaimer: I haven't tested this code at all. Only copy and paste the relevant bits from both examples mentioned above. So please expect that there some bugs in it, which needs to be fixed. But it may provide a starter.

File: CONTRIBUTING.md
Comment (jelhan) [inline CONTRIBUTING.md:31]: Do you mind adding a [script to `package.json`](https://docs.npmjs.com/cli/v8/using-npm/scripts) to simplify this? Something like this:

```json
{
  "scripts": {
    "test": "_schemas/scripts/validator.js"
  }
}
```

A user could than execute using `npm test`, which is way easier to recall.

File: package.json
Comment (jelhan) [inline package.json]: I think it would be more correct to have those as `devDependencies`. Even though it might not be relevant at all as we don't plan to publish this as a NPM package anyways.

```suggestion
  "devDependencies": {
```

File: _schemas/scripts/validator.js
Comment (jelhan) [inline _schemas/scripts/validator.js]: Currently the output in case of errors is very limited:

```
$ node ./_schemas/scripts/validator.js
Version : 1.0
Errors : 1 / 94
```

It would be great if it would log more information about the error. Especially, which test is failing and why. Maybe something like this:

```
Version : 1.0

request/resource/create/valid/post_resource.json should be valid but is invalid.
  /data/id must be string

Errors : 1 / 94
```

The needed data seems to be already available in `ko` objects:

```json
{
  version: '1.0',
  relativePath: 'request/resource/create/valid/post_resource.json',
  fullPath: '/home/main/tmp/json-api/_schemas/1.0/tests/request/resource/create/valid/post_resource.json',
  ok: false,
  errors: [
    {
      instancePath: '/data/id',
      schemaPath: 'https://jsonapi.org/schemas/spec/v1.0/draft#/definitions/resourceIdentificationNew/properties/id/type',
      keyword: 'type',
      params: [Object],
      message: 'must be string'
    }
  ]
}
```

Maybe the output of `test` command from `ajv-cli` could give some inspiration: https://ajv.js.org/packages/ajv-cli.html#test-validation-result

File: CONTRIBUTING.md
Comment (jelhan) [inline CONTRIBUTING.md]: Is it also possible to validate one test file using [`ajv-cli`](https://ajv.js.org/packages/ajv-cli.html)? I tried but wasn't able to figgure out correct options. Always ended up with this error:

```
$ ajv test --spec=draft2020 -s _schemas/1.0/schema_create_resource.json -d _schemas/1.0/tests/request/resource/create/valid/post_resource.json --valid
schema _schemas/1.0/schema_create_resource.json is invalid
error: no schema with key or ref "http://json-schema.org/draft-07/schema#"
```

**Review by jelhan — COMMENTED** [review]
Thanks a lot for adding the validation script. Looks very good!

Two topics seems to be still open:

1. What to do about existing https://jsonapi.org/schema file? Keep it for now untouched? Delete it with redirect? Delete it without redirect?
2. At which URL should we deploy the schema? Maybe https://jsonapi.org/schemas/1.0/schema.json? Do we want to provide permalinks to specific versions of the schema? E-g- https://jsonapi.org/schemas/1.0/schema-85d149dfb8834216780be0bda87ac5cf7c29338a.json?
3. Should we have a redirect alias for schema of current stable version? E.g. https://jsonapi.org/schema or https://jsonapi.org/schema.json?

@dgeb Any thoughts on that topic?

We need to update the link in FAQ as soon as we have a decision.

File: CONTRIBUTING.md
Comment (VGirol) [reply CONTRIBUTING.md]: You can not test a single file using ajv-cli for 2 reasons : 

- the json:api schema uses formats (uri, in particular) which are not included by default in the ajv validator. We need to add the ajv-format dependency to use them.
- the schemas to validate request use refenrencies to the main schema and when you use the `-s` option of the ajv-cli, you can specify only one schema. So the ajv validator doesn't know where to find the referencies...

Also, in your example, the `--spec` option must have `draft7 `as its value ;-)

I'll try to add this possibility to the script...

File: CONTRIBUTING.md
Comment (VGirol) [reply CONTRIBUTING.md]: Ok

File: CONTRIBUTING.md
Comment (VGirol) [reply CONTRIBUTING.md:31]: Ok

File: package.json
Comment (VGirol) [reply package.json]: Ok

File: _schemas/scripts/validator.js
Comment (VGirol) [reply _schemas/scripts/validator.js]: You're right.
I improved script outputs.

File: CONTRIBUTING.md
Comment (VGirol) [reply CONTRIBUTING.md]: I updated the script : now we have 2 new options : 

- `-v version` : to validate a specific version of the spec
- `-f file_path` : to validate only one test file

These 2 options can be mixed.

So, now, you can run : 
`node ./_schemas/scripts/validator.js -v 1.1 -f _schemas/1.1/tests/request/resource/create/valid/post_resource_with_lid_member.json`
or
`npm run test-schema -v 1.1 -f _schemas/1.1/tests/request/resource/create/valid/post_resource_with_lid_member.json`

**Comment by dgeb** [comment]
@jelhan good questions!

> What to do about existing https://jsonapi.org/schema file? Keep it for now untouched? Delete it with redirect? Delete it without redirect?

I see no harm in having this URL redirect to the schema for the latest version of the spec, similar to how `/format` always returns the latest version.

> At which URL should we deploy the schema? Maybe https://jsonapi.org/schemas/1.0/schema.json? 

Sure, I like this.

> Do we want to provide permalinks to specific versions of the schema? E-g- https://jsonapi.org/schemas/1.0/schema-85d149dfb8834216780be0bda87ac5cf7c29338a.json?

I'm not sure this is valuable vs. referencing github history, which has been our approach to date with the text of the spec. If we make corrections to the schema for a particular spec version, is there value in maintaining links to the old uncorrected versions?

> Should we have a redirect alias for schema of current stable version? E.g. https://jsonapi.org/schema or https://jsonapi.org/schema.json?

Yes, I would say both redirects should be supported for the sake of continuity. I'd favor `/schema.json` to become the definitive URL for the latest schema. 

@VGirol any thoughts on the above? And how are you feeling about the readiness of this PR?

**Comment by VGirol** [comment]
@dgeb, I think this PR is almost ready.
I'm waiting for @jelhan to approve (or not) one last thing : are parallel relationships supported or not ? Depending on the answer, I will have a test to keep or not.
And there is still the question about the url of the JSON schema : do I have to set up the redirect system you mention in your last comment ?

**Comment by dgeb** [comment]
> I'm waiting for @jelhan to approve (or not) one last thing : are parallel relationships supported or not ? Depending on the answer, I will have a test to keep or not.

@VGirol We are hashing out a final decision re: parallel relationships in https://github.com/json-api/json-api/issues/1361#issuecomment-1287890619

> do I have to set up the redirect system you mention in your last comment ?

This is definitely not a blocker to your work here, which has already been extensive and is much appreciated.

File: _schemas/1.0/tests/response/invalid/relationships/to_many_linkage_twice.json
Comment (jelhan) [reply _schemas/1.0/tests/response/invalid/relationships/to_many_linkage_twice.json]: > We are hashing out a final decision re: parallel relationships in [#1361 (comment)](https://github.com/json-api/json-api/issues/1361#issuecomment-1287890619)

It seems that we came to conclusion that parallel relationships are not forbidden. We consider changing that as a breaking change. At least one major implementation relies on it. As we have a strong guarantee to not add any breaking changes, we do not plan to change it.

We plan to add a recommendation that they SHOULD NOT be used. I will create a PR updating the spec as soon as possible. But as it's only a recommendation this should not affect the schema.

Would be great if you could update the schema accordingly.

Comment (VGirol) [reply _schemas/1.0/tests/response/invalid/relationships/to_many_linkage_twice.json]: Ok, i will do it next week.

Comment (VGirol) [reply _schemas/1.0/tests/response/invalid/relationships/to_many_linkage_twice.json]: @jelhan,  i've updated both schema and tests.

**Comment by auvipy** [comment]
can we check https://github.com/bowtie-json-schema/bowtie for compliance?

File: _schemas/1.0/schema.json
Comment (TehShrike) [inline _schemas/1.0/schema.json:193]: I don't know enough about JSON Schema to say why, but the latest version of ajv-cli thinks that this line makes the schema invalid:
> error: unknown format "uri" ignored in schema at path "#/definitions/linkUrl"
if I remove this one line, ajv-cli thinks the rest is fine.

Comment (VGirol) [reply _schemas/1.0/schema.json:193]: The "uri" format is provided by the ajv-formats package.
See [https://ajv.js.org/packages/ajv-formats.html](https://ajv.js.org/packages/ajv-formats.html)

**Comment by VGirol** [comment]
Is there anything that is blocking this PR? :) Should I do something? Thank you

File: _schemas/1.0/schema_update_resource.json
Comment (auvipy) [inline _schemas/1.0/schema_update_resource.json]:
```suggestion
  "$schema": "https://json-schema.org/draft/2020-12/schema",,
```

Comment (auvipy) [reply _schemas/1.0/schema_update_resource.json]: can we switch to this latest schema specs or v1.0 will have to stay with draft-7?

File: _schemas/1.0/schema_update_relationship.json
Comment (auvipy) [inline _schemas/1.0/schema_update_relationship.json]:
```suggestion
  "$schema": "https://json-schema.org/draft/2020-12/schema",
```

Comment (auvipy) [reply _schemas/1.0/schema_update_relationship.json]: same question here

File: _schemas/1.0/schema_create_resource.json
Comment (auvipy) [inline _schemas/1.0/schema_create_resource.json]:
```suggestion
  "$schema": "https://json-schema.org/draft/2020-12/schema",
```

File: _schemas/1.0/schema.json
Comment (auvipy) [inline _schemas/1.0/schema.json]:
```suggestion
  "$schema": "https://json-schema.org/draft/2020-12/schema",
```

File: .github/workflows/node.yml
Comment (auvipy) [inline .github/workflows/node.yml]:
```suggestion
      - uses: actions/checkout@v4
```

**Review by auvipy — CHANGES_REQUESTED** [review]

File: _schemas/1.0/schema_update_resource.json
Comment (VGirol) [reply _schemas/1.0/schema_update_resource.json]: IMHO we can switch to latest version.

Comment (VGirol) [reply _schemas/1.0/schema_update_resource.json]: done

File: _schemas/1.0/schema_update_relationship.json
Comment (VGirol) [reply _schemas/1.0/schema_update_relationship.json]: done

File: _schemas/1.0/schema_create_resource.json
Comment (VGirol) [reply _schemas/1.0/schema_create_resource.json]: done

File: _schemas/1.0/schema.json
Comment (VGirol) [reply _schemas/1.0/schema.json]: done

File: .github/workflows/node.yml
Comment (VGirol) [reply .github/workflows/node.yml]: done

**Comment by VGirol** [comment]
Well, i've rebased the branch to the main branch and forced-pushed the result (but i'm not sure it was a good idea...).

**Review by auvipy — CHANGES_REQUESTED** [review]
you most probably put the node_modules in gitignore?

File: _schemas/1.0/schema_create_resource.json
Comment (auvipy) [inline _schemas/1.0/schema_create_resource.json]: regarding the schema version we will be using in v1.0, we need to get input from community experts. may be using the latest schema version with v1.0 specs is OK somehow

**Review by auvipy — CHANGES_REQUESTED** [review]

**Comment by VGirol** [comment]
> you most probably put the node_modules in gitignore?

The node_modules is already put in gitignore.

File: _schemas/1.0/schema_create_resource.json
Comment (VGirol) [reply _schemas/1.0/schema_create_resource.json]: I don't see any harm in using the latest version, but it's up to you. Let's see what the community experts think about it.

**Comment by VGirol** [comment]
> > you most probably put the node_modules in gitignore?
> 
> The node_modules is already put in gitignore.

Ok, sorry, i saw what you meant : I removed the node_modules from PR.

**Comment by auvipy** [comment]
@dgeb Dan it would be great if you could have a look when you have some free time!

File: _schemas/1.0/schema_create_resource.json
Comment (bbrala) [reply _schemas/1.0/schema_create_resource.json]: By now i would say, just use the latest version

**Comment by CoolGoose** [comment]
Hello folks, any updates in this area ?

**Comment by orsinium** [comment]
Might this PR be the reason why the old link to the schema is dead?

https://jsonapi.org/schema

The link above is referenced in FAQ:

https://jsonapi.org/faq/#is-there-a-json-schema-describing-json-api

**Comment by igneus** [comment]
Yes. This PR made https://jsonapi.org/schema unavailable, not providing a (reachable) replacement.

**Comment by aillrasal195-art** [comment]
VGirol:schema-1.0

**Comment by aillrasal195-art** [comment]
https://jsonapi.org/schema

**Pull request merged**
